content/blog/eng/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline.md

```diff
@@ -0,0 +1,387 @@
+
+>ℹ️
+>
+>The complete history of my changes is tracked in the [merge request #72025](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/72025) on **GitLab**.
+
+This small **merge request** was not enough however, as `services` are handled differently by each underlying provider (**Docker** and **Kubernetes**). These providers are implemented by the **GitLab Runner** project, which provides an executable that's responsible for getting and executing `jobs` from a registered **GitLab** instance.
+
+This means that I had to add a specific logic to both the providers so that they could correctly deal with the new `variables` property defined in the `service` object.
+
+While the changes on **GitLab** itself weren't really challenging, the way **GitLab Runner** handles `environment variables` made everything a lot more difficult. Without diving deep into details, **GitLab Runner** is responsible for variables expansion at shell level, so this new feature must not break the existing workflow.
+
+The issue is better explained by [Arran Walker](https://gitlab.com/ajwalker), who reviewed my **merge request**:
+
+>💬 
+>
+>My main concern was that we want a service's variables expanded when they reference a variable that is available to the main job too.
+> 
+>For example:
+>
+>```yaml
+>variables:
+>  TOP_LEVEL_VAR: "hello"
+>
+>job:
+>  variables:
+>    JOB_LEVEL_VAR: "world"
+>  service:
+>    - image: docker:dind
+>      variables:
+>        SERVICE_LEVEL_VAR: "$TOP_LEVEL_VAR $JOB_LEVEL_VAR"
+>```
+> `SERVICE_LEVEL_VAR` should be able to reference `TOP_LEVEL_VAR` and `JOB_LEVEL_VAR`. Expanding the service variables on their own won't achieve this.
+>
+> -- <cite>https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/3158#note_740206844</cite>
+
+After few months of back and forth with **GitLab Runner**'s team, we ended up finding a solution that enabled my use case while still retaining the correct variable expansion logic, as nobody wanted to introduce a breaking-change on such a big platform.
+
+>ℹ️
+>
+>The complete history of my changes is tracked in the [merge request #3158](https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/3158) on **GitLab Runner**.
+
+My contributions were merged to **GitLab** with version **14.5** and to **GitLab Runner** with version **14.8**, making me a contributor to their platform!
+
+{{< figure src="/images/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline/contributor.png" alt="My contributor badge on GitLab" class="image featured" >}}
+
+## Our final **CI/CD** pipeline
+
+At the end of this process, we were finally able to move away from **Docker Compose** while still having a **CI/CD** pipeline which didn't need workarounds or hacks to be fully functioning. With **GitLab** 14.8, our example pipeline can now be modified as it follows:
```

Comment:
(...) that didn't need any workarounds or hacks to be fully functioning. With **GitLab** 14.8, our sample pipeline can now be modified as follows: